Sources/GRPC/Concurrency/AsyncWriter.swift

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2021, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if compiler(>=5.5)
+import NIOConcurrencyHelpers
+import NIOCore
+
+/// An asynchronous writer which forwards messages to a delegate.
+///
+/// Forwarding of messages to the delegate may be paused and resumed (by calling ``pause()``
+/// and ``resume()``, respectively). When the writer is paused calls to ``write(_:)`` may suspend.
+/// Any calls which are suspended may be continued at later (after ``resume()`` has been called).
+///
+/// The writer must also be "finished" with a final value: as for writing, calls to ``finish(_:)``
+/// may suspend if the writer has been paused.
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+internal struct AsyncWriter<Delegate: AsyncWriterDelegate> {
+  internal typealias Element = Delegate.Element
+  internal typealias End = Delegate.End
+
+  private let storage: AsyncWriterStorage<Element, End, Delegate>
+
+  internal init(maxPendingElements: Int = 16, delegate: Delegate) {
+    self.storage = AsyncWriterStorage(maxPendingElements: maxPendingElements, delegate: delegate)
+  }
+
+  /// Pauses the writer to stop value from being written to the delegate.
+  ///
+  /// This will cause subsequent calls to ``write(_:)`` to suspend until the writer is resumed.
+  internal func pause() {
+    self.storage.pause()
+  }
+
+  /// Resume the writer, allowing pending values to be written to the delegate and for suspended
+  /// calls to ``write(_:)`` to be resumed.
+  internal func resume() {
+    self.storage.resume()
+  }
+
+  /// Write an `element`.
+  ///
+  /// The call may be suspend if the writer is paused.
+  ///
+  /// Throws: ``AyncWriterError`` if the writer has already been finished or too many write tasks
+  ///   have been suspended.
+  internal func write(_ element: Element) async throws {
+    try await self.storage.write(element)
+  }
+
+  /// Write the final element
+  internal func finish(_ end: End) async throws {
+    try await self.storage.finish(end)
+  }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+fileprivate final class AsyncWriterStorage<
```

Comment:
Did you consider making this an actor?

Reply:
No -- `pause` and `resume` won't be called from asynchronous contexts.